onnxruntime/contrib_ops/cuda/collective/sharded_moe.cc

```diff
@@ -133,21 +130,30 @@ Status ShardedMoE<T>::ComputeInternal(OpKernelContext* context) const {
 
   if (moe_params.parallel_type == MoEParallelType::TP) {
     ORT_ENFORCE(moe_params.tensor_shards == nccl_->Size());
-    NCCL_RETURN_IF_ERROR(ncclGroupStart());
-    NCCL_RETURN_IF_ERROR(ncclAllReduce(reinterpret_cast<const char*>(fc2_output.get()),
-                                       reinterpret_cast<char*>(fc2_output_bc.get()), fc2_output_size / sizeof(CudaT),
-                                       GetNcclDataType(input->DataType()), ncclSum, nccl_->Comm(), Stream(context)));
-    NCCL_RETURN_IF_ERROR(ncclGroupEnd());
+
+    ORT_RETURN_IF_ERROR(FuncCustomAllReduce(nccl_,
+                                            Stream(context),
+                                            fc2_output.get(),
+                                            fc2_output_bc.get(),
+                                            static_cast<int64_t>(fc2_output_size / sizeof(CudaT)),
+                                            input->DataType(),
+                                            g_ipc_mem_res_pack_.GetIPCMemoryResourcePack()));
   }
 
   if (moe_params.parallel_type == MoEParallelType::EP) {
     size_t stride_count = moe_params.hidden_size;
     size_t stride_bytes = stride_count * sizeof(CudaT);
     int64_t total_past_rows = 0;
     int64_t total_covered_rows = 0;
+
+    static std::mutex s_mutex;
```

Comment:
> static std::mutex s_mutex; [](http://example.com/codeflow?start=3&length=27)

I think I we moved the mutex as a mutable member then it would be natural.
This serialized across all calls to this code, but I think we need to protect the data that is specific to the instance.

Reply:
Good suggestion.

Also, please move mutex inside SynchronizeExpertsStartIndex, only lock it once (no needed to lock when it is initialized (rank_to_experts_start_index_[0] != std::numeric_limits<int64_t>::min()).

Reply:
now it's lock free starting at second run